### PR TITLE
Fix #10: display system tables

### LINKED ISSUE #10: System schemas V_CATALOG and V_MONITOR not shown
Vertica provides system tables in the schemas "V_CATALOG" and "V_MONITOR" that let you monitor your database and evaluate settings of its objects.

Those tables are not shown using this driver.

DBeaver, by contrast, has an option to "Show system objects" (see attachment), which, when ticket, shows the two schemas and their tables like any other schema.
Vertica x DBeaver

### PR DESCRIPTION
**Pull request by sitingren**
Fix #10

### CODE REVIEW

File: connection.schema.json [hunk 1]
```diff
@@ -49,6 +49,11 @@
     "workload": {
       "title": "Workload",
       "type": "string"
+    },
+    "showSystemObj": {
```
Comment (DMickens) [inline connection.schema.json]: nit: showSystemObjs

I don't have a preference either way, I'm just curious about the decision to default to false. Was there a specific reason for it?

File: src/ls/queries.ts [hunk 2]
```diff
@@ -3,8 +3,9 @@ import escapeTableName from '../escape-table';
 import { IBaseQueries, ContextValue } from '@sqltools/types';
 
 const describeTable: IBaseQueries['describeTable'] = queryFactory`
-SELECT * FROM
-  ${p => p.isView ? `V_CATALOG.VIEW_COLUMNS` : `V_CATALOG.COLUMNS`}
+SELECT * FROM ${p => p.isView ? `V_CATALOG.VIEW_COLUMNS` : `(SELECT * FROM V_CATALOG.COLUMNS
```
Comment (DMickens) [inline src/ls/queries.ts]: I think we are going to run into compatibility issues here with different versions of Vertica. Because V_CATALOG.COLUMNS now has a table_namespace column and V_CATALOG.SYSTEM_COLUMNS does not, this union won't succeed against the soon-to-be released version of Vertica. The new column also isn't at the end of the table, so it's not as clean as just doing a select *, null column_set_using, null is_identity, null table_namespace, it would become:

SELECT * FROM (SELECT * FROM V_CATALOG.COLUMNS UNION SELECT table_id, **null table_namespace**, table_schema, table_name, is_system_table, column_id, column_name, data_type, data_type_id, data_type_length, character_maximum_length, numeric_precision, numeric_scale, datetime_precision, interval_precision, ordinal_position, is_nullable, column_default, null column_set_using, null is_identity FROM V_CATALOG.SYSTEM_COLUMNS) AS U;

I can talk to Upgrade team to see if there is something that we can do to resolve this, but as it is now, with UNION in Vertica there is a good chance we will run into compatibility issues.

File: src/ls/queries.ts [hunk 3]
```diff
@@ -34,6 +35,35 @@ ORDER BY
...
+  C.CHARACTER_MAXIMUM_LENGTH::INT AS size,
+  '${p => p.database}' AS database,
+  C.TABLE_SCHEMA AS schema,
+  ${p => p.isView ? `NULL` : `C.COLUMN_DEFAULT`} AS "defaultValue",
+  ${p => p.isView ? `NULL` : `C.IS_NULLABLE`} AS "isNullable",
+  false as "isPk",
+  false as "isFk"
+FROM
+  ${p => p.isView ? `V_CATALOG.VIEW_COLUMNS` : `(
+    SELECT COLUMN_NAME,TABLE_NAME,TABLE_SCHEMA,DATA_TYPE,CHARACTER_MAXIMUM_LENGTH,COLUMN_DEFAULT,IS_NULLABLE,ORDINAL_POSITION
```
Comment (DMickens) [inline src/ls/queries.ts]: nit: space after each comma makes it easier on the eyes to parse

File: src/ls/queries.ts [hunk 4]
```diff
@@ -77,6 +107,45 @@ ORDER BY
...
+const searchTablesAll: IBaseQueries['searchTables'] = queryFactory`
+SELECT U.*, D.database_name AS database
+  FROM (
+      SELECT
+          T.TABLE_NAME AS label,
+          '${ContextValue.TABLE}' AS type,
+          T.TABLE_SCHEMA AS schema,
+          FALSE AS "isView",
+          'table' AS description,
+          ('"' || T.TABLE_SCHEMA || '"."' || T.TABLE_NAME || '"') as detail
```
Comment (DMickens) [inline src/ls/queries.ts:118]: I don't quite understand this line. What is this for?

File: src/ls/queries.ts (hunk 2 above)
Comment (DMickens) [reply src/ls/queries.ts]: Another option is to list out everything explicitly except table_namespace from V_CATALOG.COLUMNS which would avoid the compatibility issues in the short term, but then we wouldn't have the namespace context to provide to users

File: src/ls/queries.ts [hunk 5]
```diff
@@ -177,17 +333,35 @@ WHERE
   IS_SYSTEM_SCHEMA = 'f'
 ORDER BY schema_name
 `;
-// OR schema_name in ('v_catalog', 'v_monitor')
+
+// hide v_internal schema
```
Comment (DMickens) [inline src/ls/queries.ts]: good catch

**Review by DMickens — COMMENTED** [review]
Looks generally good overall. I have concerns about the namespace issue, but I'm trying to get some information about our options there. We probably want to wait on merging, or at least wait to publish until we have some clarity there. 

Other than that I just had a few small other comments. 

I'm glad I wasn't the one to have to put together these queries 😅

File: connection.schema.json (hunk 1 above)
Comment (williammatherjones) [reply connection.schema.json]: It should default to false because the normal SQLTools drivers do not include system objects.  

I agree it should be plural.   How about showSystemTables?  showSystemObjects?  I find showSystemObjs to be hard to parse.   I prefer showSystemTables.  I realize this includes the tables and the schemas they are in, but I think that is clear.

Comment (DMickens) [reply connection.schema.json]: It's also for views in some cases though, so it's schemas, views, and tables. Objects might object-ively be a better option😄

File: src/ls/queries.ts (hunk 2 above)
Comment (williammatherjones) [reply src/ls/queries.ts]: Why do we need to include the table_namespace column?   The server implementation Yogesh did should return the _namespace_._schema_ format for the _table_schema_ column.  SQLTools does not know what to do with namespaces and does not need the value.    As long as the schema names are qualified with the namespace names the integration should work.    Siting was going to test with an Eon database with namespaces defined just to make sure.

Comment (williammatherjones) [reply src/ls/queries.ts]: V_CATALOG.COLUMNS and V_CATALOG.VIEW_COLUMNS is not in the list of tables I identified as catalog tables that were used by the drivers at the time we came up with the approach:  https://confluence.verticacorp.com/display/DEV/Namespace+Impact+on+Client+Drivers

Comment (DMickens) [reply src/ls/queries.ts]: If we don't want to include table_namespace column at all then we could go with my previous suggestion to list out everything except for table_namespace. I'm not sure what the implications are for that in terms of user experience. If that was removed from this union query what would change?

File: connection.schema.json (hunk 1 above)
Comment (williammatherjones) [reply connection.schema.json]: Then how about showSystemObjects?   Screens are big these days and there's less need to abbreviate...

File: src/ls/queries.ts (hunk 4 above)
Comment (sitingren) [reply src/ls/queries.ts:118]: This is a piece of info that the sqltools framework needs. Not sure how sqltools use it.

File: connection.schema.json (hunk 1 above)
Comment (sitingren) [reply connection.schema.json]: @DMickens You cannot create a view under a system schema, so there is no "system views". Views link with system tables will be created under public schema by default.

@williammatherjones I'll update the variable name.

File: src/ls/queries.ts (hunk 2 above)
Comment (sitingren) [reply src/ls/queries.ts]: I have updated this query with a new way to determine the correct table: using javascript to filter v_catalog and v_monitor schema.

In 24.2 server, 'table_namespace' column is added into multiple system tables. If its value is `default_namespace`, then 'table_schema' column is just the schema; otherwise 'table_schema' column is namespace.schema.

And as a result, user can see schema list as below:
image

File: src/ls/queries.ts (hunk 3 above)
Comment (sitingren) [reply src/ls/queries.ts]: Done

File: src/ls/queries.ts [hunk 6]
```diff
@@ -177,17 +332,35 @@ WHERE
...
+const fetchSchemasAll: IBaseQueries['fetchSchemas'] = queryFactory`
+SELECT
+  schema_name AS label,
+  schema_name AS schema,
+  '${ContextValue.SCHEMA}' as "type",
+  'group-by-ref-type' as "iconId",
+  '${p => p.database}' as database
+FROM V_CATALOG.SCHEMATA
+WHERE
+  schema_name NOT IN ('v_internal', 'v_license')
```
Comment (sitingren) [inline src/ls/queries.ts:346]: There is a new system table `v_license` added in v24.2. Similar to v_internal, no table can be listed in V_CATALOG.SYSTEM_TABLES. I'm not sure whether we should hide it as well.

File: src/ls/queries.ts (hunk 2 above)
Comment (williammatherjones) [reply src/ls/queries.ts]: That's the correct behavior.

File: src/ls/queries.ts (hunk 4 above)
Comment (williammatherjones) [reply src/ls/queries.ts:118]: I think they need the fully qualified table name. Since the schema name includes the namespace if appropriate, this will be a name like ns2.s1.t1, or for the non-namespace case, s1.t1.

File: src/ls/queries.ts (hunk 6 above)
Comment (williammatherjones) [reply src/ls/queries.ts:346]: 'v_catalog' and 'v_monitor' are the only documented system schemas.   Let's leave `v_license` out.

**Review by williammatherjones — APPROVED** [review]
Looks good.  I think this is ready.
